microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricsRegistrationService.java

```diff
@@ -70,22 +75,50 @@ static void install(OperationContext context, List<String> exposedSubsystems, St
         Supplier<ModelControllerClientFactory> modelControllerClientFactory = serviceBuilder.requires(context.getCapabilityServiceName(CLIENT_FACTORY_CAPABILITY, ModelControllerClientFactory.class));
         Supplier<Executor> managementExecutor = serviceBuilder.requires(context.getCapabilityServiceName(MANAGEMENT_EXECUTOR, Executor.class));
         serviceBuilder.requires(CONFIG_PROVIDER);
-        MetricsRegistrationService service = new MetricsRegistrationService(rootResourceRegistration, rootResource, modelControllerClientFactory, managementExecutor, exposedSubsystems, prefix);
+        Supplier<ExecutorService> executorService = Services.requireServerExecutor(serviceBuilder);
+        MetricsRegistrationService service = new MetricsRegistrationService(rootResourceRegistration, rootResource, modelControllerClientFactory, managementExecutor, executorService, exposedSubsystems, prefix);
         serviceBuilder.setInstance(service)
                 .install();
     }
 
-    public MetricsRegistrationService(ImmutableManagementResourceRegistration rootResourceRegistration, Resource rootResource, Supplier<ModelControllerClientFactory> modelControllerClientFactory, Supplier<Executor> managementExecutor, List<String> exposedSubsystems, String globalPrefix) {
+    public MetricsRegistrationService(ImmutableManagementResourceRegistration rootResourceRegistration, Resource rootResource, Supplier<ModelControllerClientFactory> modelControllerClientFactory, Supplier<Executor> managementExecutor, Supplier<ExecutorService> executorService, List<String> exposedSubsystems, String globalPrefix) {
         this.rootResourceRegistration = rootResourceRegistration;
         this.rootResource = rootResource;
         this.modelControllerClientFactory = modelControllerClientFactory;
         this.managementExecutor = managementExecutor;
+        this.executorService = executorService;
         this.exposedSubsystems = exposedSubsystems;
         this.globalPrefix = globalPrefix;
     }
 
     @Override
     public void start(StartContext context) throws StartException {
+        final Runnable task = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    doStart(context);
+                } catch (StartException e) {
+                    context.failed(e);
+                }
+            }
+        };
+        try {
+            executorService.get().submit(task);
+        } catch (RejectedExecutionException e) {
+            task.run();
+        } finally {
+            context.complete();
+        }
+
+    }
+
+    private void doStart(StartContext context) throws StartException {
+        final ServiceContainer serviceContainer = context.getController().getServiceContainer();
+        try {
+            serviceContainer.awaitStability();
```

Comment:
In theory you might get a false positive re stability, since some OSHs may not have made their MSC changes when this task runs. So it's stable for a millisecond and then it's destabilized again.

When the OperationContext awaits stability it knows all the Stage.RUNTIMES OSHs have completed execution.

Reply:
If this proves troublesome we can go back to the Stage.VERIFY approach.